port/raspberrypi/rp2xxx/src/hal/i2c_slave.zig

```diff
@@ -0,0 +1,312 @@
+        .IC_10BITADDR_SLAVE = .ADDR_7BITS,
+        .IC_10BITADDR_MASTER = .ADDR_7BITS,
+        .STOP_DET_IFADDRESSED = .DISABLED,
+        .STOP_DET_IF_MASTER_ACTIVE = 0,
+        .padding = 0,
+    });
+
+    // Be careful when setting the mask.  Due to vagrancies of the RP2040.svd file
+    // the values are the opposite of what you might expect.  When setting the mask:
+    //  .ENABLED    is 0 (masked)   which means the interrupt is disabled, and
+    //  .DISABLED   is 1 (unmasked) which means the interrupt is enabled.
+
+    self.regs.IC_INTR_MASK.write(.{
+        .M_RD_REQ = .DISABLED,
+        .M_RX_FULL = .DISABLED,
+        .M_STOP_DET = .DISABLED,
+        .M_RESTART_DET = .DISABLED,
+        .M_TX_ABRT = .DISABLED,
+
+        .M_GEN_CALL = .ENABLED,
+        .M_START_DET = .ENABLED,
+        .M_ACTIVITY = .ENABLED,
+        .M_RX_DONE = .ENABLED,
+        .M_TX_EMPTY = .ENABLED,
+        .M_TX_OVER = .ENABLED,
+        .M_RX_OVER = .ENABLED,
+        .M_RX_UNDER = .ENABLED,
+        .padding = 0,
+    });
+
+    // ### TODO ### Set up the ISR at runtime
+    // Ideally, MicroZig would have a way to set up the ISR at runtime,
+    // and we could make a call like:
+    //    `irq.set_handler( self.i2c, if (i2c == 0) isr0 else isr1);`
+
+    // For now, you'll have to set up the ISR manually in your main program.
+    // by adding something like the following to your main program:
+    //
+    // pub const microzig_options = microzig.Options
+    //   {
+    //     .interrupts = .{ .I2C0_IRQ = .{ .c = i2c_slave.isr0 }, .I2C1_IRQ = .{ .c = i2c_slave.isr1 } },
+    //   } ;
+
+    self.enable();
+    irq.enable(.I2C1_IRQ);
+}
+
+/// Close the I2C slave driver
+///
+/// ## Parameters
+///
+/// * `self` - The I2C slave driver to close
+///
+pub fn close(self: *Self) void {
+    self.disable();
+}
+
+/// Disable the I2C slave
+inline fn disable(self: *Self) void {
+    self.regs.IC_ENABLE.write(.{
+        .ENABLE = .DISABLED,
+        .ABORT = .DISABLE,
+        .TX_CMD_BLOCK = .NOT_BLOCKED,
+        .padding = 0,
+    });
+}
+
+/// Enable the I2C slave
+inline fn enable(self: *Self) void {
+    self.regs.IC_ENABLE.write(.{
+        .ENABLE = .ENABLED,
+        .ABORT = .DISABLE,
+        .TX_CMD_BLOCK = .NOT_BLOCKED,
+        .padding = 0,
+    });
+}
+
+/// Set the I2C slave address
+///
+/// ## Parameters
+///
+/// * `self` - The I2C0 or I2C1 slave driver to use
+/// * `addr` - The I2C slave address to use
+///
+pub fn set_slave_address(self: *Self, addr: u7) void {
+    self.disable();
+    self.regs.IC_SAR.write(.{ .IC_SAR = @enumFromInt(addr), .padding = 0 });
+    self.enable();
+}
+
+/// Common ISR for the I2C0 and I2C1 slave driver
+///
+/// ## Parameters
+///
+/// * `self` - The I2C0 or I2C1 slave driver to use
+///
+fn isr_common(self: *Self) void {
+
+    // Save the interrupt status and clear it
+
+    const interruptStatus = self.regs.IC_RAW_INTR_STAT.read();
+    _ = self.regs.IC_CLR_INTR.read();
```

Comment:
Is it OK to clear it here? Could this interrupt be jumped to again if another one comes it before we return? If so we should `defer` this.

Reply:
I believe it is OK.  My understanding is that only interrupts of higher priority can interrupt an in progress interrupt.  Doing it this way, if something significant happens after we clear the bits, the appropriate bit will be set and the interrupt will be re-triggered once we return.

Reply:
Right. That makes sense.